### PR TITLE
CI C++17

### PR DESCRIPTION
**Pull request by nilsnolde**
start with #3877 , still wait for more feedback. more a quick test.

### CODE REVIEW

**Comment by nilsnolde** [comment]
hm, of course some `-WError` kicks in. 

on linux, seems like the `date` library has an `uncaught_exception` warning which some of our source files catch onto. it was apparently [fixed for windows](https://github.com/HowardHinnant/date/commit/de6a03d337dcd617a121254f700979086889c0a4), but seems to be an issue for unix as well, as [someone says](https://github.com/HowardHinnant/date/issues/338#issuecomment-485915937) it should say `if __cplusplus > 201703` [here](https://github.com/HowardHinnant/date/blob/8a563041fa65146531249ed557a6839809488521/include/date/date.h#L136), at least that's what locally works. worth warming up probably and PR upstream, right?

on macos, there's a few others in meili mostly apparently, mostly `error: 'iterator<std::forward_iterator_tag, const valhalla::meili::Label>' is deprecated [-Werror,-Wdeprecated-declarations]`. at least that's in our control.. but also some in libosmium with `error: 'result_of<osmium::io::detail::SerializeBlob ()>' is deprecated [-Werror,-Wdeprecated-declarations]`, which doesn't seem to be reported anywhere upstream. kinda weird, considering c++17 is in wide use, I'd have expected someone else to build with `-Werror`. but then again, seems to be macos(/xcode?) issue, builds fine on linux locally with the quick fix to overcome the `date` issue, hope CI will agree.

**Comment by nilsnolde** [comment]
ah good, the libosmium issue was already fixed in https://github.com/osmcode/libosmium/commit/e933414c191f2209e9869247d2b2eede02f77c86. and was release with [2.17.2](https://github.com/osmcode/libosmium/releases/tag/v2.17.2). we're on 2.16.0. so I guess moving up the version should solve that. might as well go all the way to 2.18.0 (lots of deprecated stuff was removed there, but 2.17.x should work for us).

**Comment by nilsnolde** [comment]
Back on my office machine, it also failed with gcc 12.2.0 with the same std::iterator deprecations when building the tests. Strange it works on CI Linux machines. There's a nice post about the why: https://www.fluentcpp.com/2018/05/08/std-iterator-deprecated/. I (tentatively) fixed it quickly and I could build the test locally.

Also upgraded to the latest libosmium release which went pretty smooth.

File: valhalla/meili/routing.h [hunk 1]
```diff
@@ -305,8 +305,14 @@ find_shortest_path(baldr::GraphReader& reader,
...
 
 // Route path iterator. Methods to assist recovering route paths from Labels.
-class RoutePathIterator : public std::iterator<std::forward_iterator_tag, const Label> {
+class RoutePathIterator {
 public:
+  using iterator_category = std::forward_iterator_tag;
+  using value_type = const Label;
+  using difference_type = const std::ptrdiff_t;
+  using pointer = const Label*;
+  using reference = const Label&;
```
Comment (nilsnolde) [inline valhalla/meili/routing.h:314]: basically what `std::iterator` was doing under the hood (also the only thing it was really doing).

Comment (nilsnolde) [reply valhalla/meili/routing.h:314]: let's reason over the constness together..

**Review by kevinkreiser — APPROVED** [review]

**Pull request merged**